lib/clib/CCorresLemmas.thy

```diff
@@ -1040,119 +1040,211 @@ lemma ccorres_Guard_True_Seq:
+      apply (drule_tac x=r in meta_spec)
+      apply (rule_tac Q'="{s' \<in> G'. cond_xf s' \<noteq> 0 \<longrightarrow> s' \<in> C'}" in conseqPost)
+        apply fastforce
+       apply fastforce
+      apply fastforce
+     apply (rule hoarep_conj_lift_pre_fix)
+      apply (simp add: Collect_mono conseq_under_new_pre)
+     apply (simp add: Collect_mono conseq_under_new_pre)
+    apply (insert setter_inv_cond)
+    apply (drule_tac x=s in meta_spec)
+    apply (drule_tac x=r in meta_spec)
+    apply (simp add: imp_conjR)
+    apply (rule hoarep_conj_lift_pre_fix)
+     apply (simp add: Collect_mono conseq_under_new_pre)
+    apply (rule_tac Q'="{s'. C r s \<noteq> None \<and> the (C r s) = (cond_xf s' \<noteq> 0)}"
+                 in conseqPost[rotated])
+      apply fastforce
+     apply fastforce
+    apply (simp add: Collect_mono conseq_under_new_pre)
+    done
+
   show ?thesis
     apply (clarsimp simp: ccorres_underlying_def)
     apply (rename_tac s s' n xstate)
-    apply (frule (1) exec_handlers_use_hoare_nothrow_hoarep)
-     apply (rule_tac R="G' \<inter> {s'. s' \<in> {t. cond_xf t \<noteq> 0} \<longrightarrow> s' \<in> C'}" in HoarePartial.Seq)
-      apply (fastforce intro: setter_inv_cond)
-     apply (fastforce intro: While_inv_from_body_setter setter_inv_cond body_inv)
-    apply clarsimp
-    apply (frule (1) intermediate_Normal_state)
-     apply (fastforce intro: setter_inv_cond)
+    apply (frule_tac R'="{s'. s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')}"
+                 and Q'="{s'. \<exists>rv s. s' \<in> G' \<and> (s, s') \<in> srel \<and> G rv s \<and> rrel rv (xf s')}"
+                  in exec_handlers_use_hoare_nothrow_hoarep)
+      apply fastforce
+     apply (rule HoarePartial.Seq)
+      apply (erule setter_hoarep)
+     apply (rule conseqPre)
+      apply (rule ccorres_While_Normal_helper)
+       apply (fastforce intro!: setter_hoarep hoarep_ex_lift)
+      apply (intro hoarep_ex_pre, rename_tac rv new_s)
+      apply (insert setter_inv_cond)[1]
+      apply (drule_tac x=new_s in meta_spec)
+      apply (drule_tac x=rv in meta_spec)
+      apply (insert body_ccorres)[1]
+      apply (drule_tac x=rv in meta_spec)
+      apply (insert body_inv(2))[1]
+      apply (drule_tac x=new_s in meta_spec)
+      apply (drule_tac x=rv in meta_spec)
+      apply (frule_tac s=new_s in ccorres_to_vcg_with_prop')
+        using nf apply fastforce
+       using body_inv apply fastforce
+      apply (rule_tac Q'="{s'. s' \<in> G' \<and> (\<exists>rv s. (s, s') \<in> srel \<and> rrel rv (xf s') \<and> G rv s)}"
+                   in conseqPost)
+        apply (rule hoarep_conj_lift_pre_fix)
+         apply fastforce
+        apply (simp add: conseq_under_new_pre)
+       apply fastforce
+      apply fastforce
+     apply fastforce
+    apply (case_tac xstate; clarsimp)
+    apply (frule intermediate_Normal_state[OF _ _ setter_hoarep])
+      apply fastforce
+     apply fastforce
```

Comment:
`; fastforce`

Reply:
That spins too. Here I can do `blah; assumption?` and then just one `fastforce`, which is again just one line better.

Reply:
Might also be the order again, so `(.., fastforce+)[1]` might have a chance.